PCF-ADR-0004-codeflare-test-strategy.md

```diff
@@ -0,0 +1,266 @@
+  * Checkout the branch to be tested, e.g., the PR feature branch, i.e. `git clone --branch mytag0.1 --depth 1 https://example.com/my/repo.git`
+  * Build the CodeFlare operator container image at HEAD revision
+  * Push the CodeFlare operator container image into a test container image registry (resp. the OpenShift cluster internal image registry)
+  * Build the OLM bundle image
+  * Push the OLM bundle image into a test container image registry (resp. the OpenShift cluster internal image registry)
+4. Build the OLM index image:
+  * Build the OLM index/catalog image with OPM from the latest Operator Hub community catalog (resp. OpenShift catalog)
+
+> **Note**
+> the above should be done according to https://github.com/operator-framework/operator-sdk/issues/5832 when available.
+
+> **Note**
+> the channel must be adapted according to https://github.com/project-codeflare/codeflare-operator/issues/126.
+
+  * Push the OLM index/catalog image that’s been built at the previous step to the test container image registry (resp. The OpenShift cluster internal container image registry) 
+
+
+The test cases can now be executed, with a **cluster admin role**, in sequence, as the upgrade test case depends on the installation one.
+
+##### Installation of the CodeFlare Operator with OLM
+
+###### Description
+
+Install the latest released version of the CodeFlare operator using OLM, and assert successful deployment of the operator. Run a smoke test, to make sure the current version of the operator is working as expected. 
+
+###### Scenario
+
+1. Create a codeflare-system test Namespace (resp. Project)
+2. Create a CatalogSource resource pointing to the latest release index image
+3. Create a Subscription resource, that points to the CatalogSource resource created at the previous step
+4. Run a smoke test, e.g. by running one of the e2e test cases, to make sure the operator current version is working correctly. 
+
+##### Upgrade of the CodeFlare Operator installed via OLM
+
+###### Description
+
+Add a newer version of the CodeFlare operator bundle, replacing the latest released version from the channel subscribed to during the Installation of the CodeFlare Operator with OLM test case. Assert successful upgrade of the operator to the newer version, and run a smoke test, to make sure that newer version is working as expected. 
+
+###### Scenario
+
+> **Note**
+> Note: Make sure to execute the Installation of the CodeFlare Operator with OLM test case first.
+
+1. In the same codeflare-system Namespace (resp. Project), update the existing test-catalog CatalogSource resource, by pointing to the newly built OLM index image
+2. Assert a new ClusterServiceVersion resource has been created for the newer operator version, and eventually reaches the CSVPhaseSucceeded phase
+3. Run a smoke test, e.g., by running one of the e2e test cases, to make sure the operator current version is working correctly.
+
+## Open Questions
+
+1. Add observability test cases
```

Comment:
I'd vote for us to include additional ADRS in the future / update this one later. Perhaps in the short-term we just create github issues to follow up on?

Reply:
+1

Reply:
@sutaakar once the issues are created could you update the ADR such that each open question links to the corresponding issue? Once that's done, I'm good to merge.

Reply:
@anishasthana I have added issue for observability test cases.
About integration tests (ODH), there are some basic tests in https://github.com/opendatahub-io/distributed-workloads/tree/main/tests . What is our approach to those tests, should we leverage/expand those tests?

Reply:
I think we should expand/replace those tests as desired. We should not feel bound to reusing the existing bash-based tests.

Reply:
@anishasthana Added a link to task for adding/revising integration tests.